controllers/packetmachine_controller.go

```diff
@@ -146,85 +170,85 @@ func (r *PacketMachineReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		err = r.reconcileDelete(ctx, machineScope)
 		return ctrl.Result{}, err
 	}
-
 	return r.reconcile(ctx, machineScope)
 }
 
 func (r *PacketMachineReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
 	log := ctrl.LoggerFrom(ctx)
 
-	clusterToObjectFunc, err := util.ClusterToObjectsMapper(r.Client, &infrav1.PacketMachineList{}, mgr.GetScheme())
+	clusterToPacketMachines, err := util.ClusterToTypedObjectsMapper(mgr.GetClient(), &infrav1.PacketMachineList{}, mgr.GetScheme())
 	if err != nil {
 		return fmt.Errorf("failed to create mapper for Cluster to PacketMachines: %w", err)
 	}
 
-	return ctrl.NewControllerManagedBy(mgr).
-		WithOptions(options).
+	err = ctrl.NewControllerManagedBy(mgr).
 		For(&infrav1.PacketMachine{}).
+		WithOptions(options).
 		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(log, r.WatchFilterValue)).
 		Watches(
-			&source.Kind{Type: &clusterv1.Machine{}},
+			&clusterv1.Machine{},
 			handler.EnqueueRequestsFromMapFunc(util.MachineToInfrastructureMapFunc(infrav1.GroupVersion.WithKind("PacketMachine"))),
 		).
 		Watches(
-			&source.Kind{Type: &infrav1.PacketCluster{}},
-			handler.EnqueueRequestsFromMapFunc(r.PacketClusterToPacketMachines(ctx)),
+			&infrav1.PacketCluster{},
+			handler.EnqueueRequestsFromMapFunc(r.PacketClusterToPacketMachines),
 		).
 		Watches(
-			&source.Kind{Type: &clusterv1.Cluster{}},
-			handler.EnqueueRequestsFromMapFunc(clusterToObjectFunc),
-			builder.WithPredicates(predicates.ClusterUnpausedAndInfrastructureReady(log)),
-		).
-		Complete(r)
+			&clusterv1.Cluster{},
+			handler.EnqueueRequestsFromMapFunc(clusterToPacketMachines),
+			builder.WithPredicates(
+				predicates.ClusterUnpausedAndInfrastructureReady(log),
+			),
+		).Complete(r)
+	if err != nil {
+		return fmt.Errorf("failed setting up with a controller manager: %w", err)
+	}
+	return nil
 }
 
 // PacketClusterToPacketMachines is a handler.ToRequestsFunc to be used to enqeue requests for reconciliation
 // of PacketMachines.
-func (r *PacketMachineReconciler) PacketClusterToPacketMachines(ctx context.Context) handler.MapFunc {
+func (r *PacketMachineReconciler) PacketClusterToPacketMachines(ctx context.Context, o client.Object) []ctrl.Request {
 	log := ctrl.LoggerFrom(ctx)
-	return func(o client.Object) []ctrl.Request {
-		c, ok := o.(*infrav1.PacketCluster)
-		if !ok {
-			log.Error(fmt.Errorf("expected a PacketCluster but got a %T", o), "failed to get PacketMachine for PacketCluster") //nolint:goerr113
-			return nil
-		}
-
-		log = log.WithValues("PacketCluster", c.Name, "Namespace", c.Namespace)
-
-		// Don't handle deleted PacketClusters
-		if !c.ObjectMeta.DeletionTimestamp.IsZero() {
-			log.V(4).Info("PacketCluster has a deletion timestamp, skipping mapping.")
-			return nil
-		}
+	result := []ctrl.Request{}
+	c, ok := o.(*infrav1.PacketCluster)
+	if !ok {
+		panic(fmt.Sprintf("expected a PacketCluster but got a %T", o))
```

Comment:
panic in a controller can strand resources that would be managed in the next reconciliation loop. I suspect we'll want to log an error and return nil instead.

Reply:
https://github.com/kubernetes-sigs/cluster-api/pull/3735/commits/9eaa3f543b34f6b2d024d2dc749dd2acf39a2137

So here's the upstream patch that switches to it. I'm a bit confused to the why myself.

Reply:
Made this not a panic.